listenbrainz_spark/recommendations/candidate_sets.py

```diff
@@ -0,0 +1,363 @@
+            .format(config.STARTING_YEAR, "%02d" % config.STARTING_MONTH, config.ENDING_YEAR, "%02d" % config.ENDING_MONTH))
+
+    artists_relation_df = None
+    try:
+        path = '{}/data/listenbrainz/similar_artists/artist_artist_relations.parquet'.format(config.HDFS_CLUSTER_URI)
+        artists_relation_df = listenbrainz_spark.sql_context.read.parquet(path)
+    except AnalysisException as err:
+        logging.error('Cannot read artist-artist relation from HDFS: {} \n {}. Aborting...'.format(type(err).__name__,
+            str(err)))
+        sys.exit(-1)
+    except Exception as err:
+        logging.error('An error occured while fecthing artist-artist relation: {} \n {}. Aborting...'.format(
+            type(err).__name__, str(err)))
+        sys.exit(-1)
+    if not artists_relation_df:
+        raise SystemExit("Parquet file conatining artist-artist relation is missing from HDFS. Aborting...")
+
+    try:
+        path = os.path.join(config.HDFS_CLUSTER_URI, 'data', 'listenbrainz', 'recommendation-engine', 'dataframes')
+        recordings_df = listenbrainz_spark.sql_context.read.parquet(path + '/recordings_df.parquet')
+        users_df = listenbrainz_spark.sql_context.read.parquet(path + '/users_df.parquet')
+    except AnalysisException as err:
+        logging.error('Cannot read parquet files from HDFS: {} \n {}'.format(type(err).__name__,str(err)))
+        sys.exit(-1)
+    except Exception as err:
+        logging.error('An error occured while fetching parquets: {} \n {}. Aborting...'.format(type(err).__name__,
+            str(err)), exc_info=True)
+        sys.exit(-1)
+
+    logging.info('Registering Dataframes...')
+    try:
+        listens_df.createOrReplaceTempView('listens_df')
+        recordings_df.createOrReplaceTempView('recording')
+        users_df.createOrReplaceTempView('user')
+        artists_relation_df.createOrReplaceTempView('artists_relation')
+    except AnalysisException as err:
+        logging.error('Cannot register dataframe: {} \n {}. Aborting...'.format(type(err).__name__, str(err)))
+        sys.exit(-1)
+    except Exception as err:
+        logging.error('An error occured while registering dataframe: {} \n {}. Aborting...'.format(type(err).__name__,
+            str(err)), exc_info=True)
+        sys.exit(-1)
+    logging.info('Files fectched from HDFS and dataframes registered in {}s'.format('{:.2f}'.format(time() - ti)))
+
+    path = os.path.join(os.path.dirname(os.path.abspath(__file__)),'recommendation-metadata.json')
+    user_data = defaultdict(dict)
+    with open(path) as f:
+        recommendation_metadata = json.load(f)
+        similar_artists_candidate_set_df = None
+        top_artists_candidate_set_df = None
+        for user_name in recommendation_metadata['user_name']:
+            top_artists = ()
+            similar_artists = ()
+            top_artists_with_collab = ()
+            try:
```

Comment:
This try block is probably a bit too large, debugging problems in here would not be easy, we should break it into smaller try catch blocks with different log messages.